### DISCUSSION TITLE
How to disable compiled cache on a relationship?

### DISCUSSION THREAD

**Discussion by dkbarn** [sqlalchemy/sqlalchemy#10895] [Usage Questions]
I have a `relationship()` that when evaluated is triggering the following warning:

> SAWarning: Loader depth for query is excessively deep; caching will be disabled for additional loaders.   For recursive eager loaders consider using the recursion_depth feature.  Use the compiled_cache=None execution option to skip this warning.

Elsewhere in my code when a query triggers this warning, I address it by manually passing `execution_options(compiled_cache=None)` onto the select statement. But I don't see any way of passing execution options into a relationship.

How can I disable the cache for the evaluation of a specific relationship which involves a deeply nested query?

**Comment by zzzeek** [comment]
try sending it to the execute call instead:


    result = session.execute(stmt, execution_options={"compiled_cache": None})

though you'd be much better off just silencing the warning directly.

**Reply by dkbarn** [reply]
This is for a relationship though, the query happens as soon as you reference `mapped_obj.relationship_attr` so there's not a call to `session.execute` that I can tap into.

I have this at the start of my application:
`warnings.filterwarnings("error", category=sqlalchemy.exc.SAWarning)`
specifically because I am in the phase of migrating from 1.4 to 2.0 and it's important that I have visibility on every warning SA is issuing to ensure my application operates correctly against the new version, so I would rather not have to silence a warning directly.

It seems less than ideal for SA to issue a warning about a situation but provide no way to address that warning short of silencing all warnings.

**Reply by zzzeek** [reply]
this was also addressed in #10230 and #10231 but these would not apply to a "lazyload".  came up also in #10579.  in #10579 I suggested we at least provide an explicit option (likely a parameter on ``relationship``) that indicates what depth of caching is desired, after which the caching turns off.   using this parameter would disable the warning entirely.   I did not open a new issue for this feature at the time but that still should be done.

it's not that easy to trigger this off of a lazyload IIRC how this works, unless you have loader options that come off of lazyload, like `lazyload(A.bs).selectinload(B.cs)` kind of thing?    The issue is that the very long chain of loader options creates a cache key that is very large and not really worth the memory overhead in generating it.    There are usually ways to reorganize queries so that this condition doesn't occur in the first place - otherwise there would be no point to having the warning, we'd have to silently just not cache the too-long set of options.

If it turns out there's just a large, common class of queries that are impossible to reorganize to work around this limitation, then we'll just take the warning out.   

We also should open an issue for the feature suggested in #10579.

**Reply by dkbarn** [reply]
I've been digging into my particular case a bit more to see if I can address the warning by changing the query to not be so deep. I'm actually unclear on why this particular query is triggering the warning.

The relationship in question is defined as:
```python
last_online = (
    select(AccountOnline.id)
        .where(AccountOnline.account_id == Account.id)
        .order_by(AccountOnline.timestamp.desc())
        .limit(1)
        .correlate(Account)
        .as_scalar()
)

Account.last_online = relationship(
    "AccountOnline",
    primaryjoin=and_(
        AccountOnline.id == last_online,
        AccountOnline.account_id == Account.id,
    ),
    uselist=False,
    viewonly=True,
)
```

I set a breakpoint here: https://github.com/sqlalchemy/sqlalchemy/blob/822703b7c67fe59ad6d2bf2bca7002cd7153a85d/lib/sqlalchemy/orm/context.py#L520

This was the state at that point:
```
>>> print(statement)
SELECT account_online.id AS account_online_id, account_online.account_id AS account_online_account_id, account_online.timestamp AS account_online_timestamp, account_online.timestamp_accuracy AS account_online_timestamp_accuracy, account_online.reason AS account_online_reason, account_online.changes AS account_online_changes 
FROM account_online 
WHERE account_online.id = (SELECT account_online.id 
FROM account_online 
WHERE account_online.account_id = :param_1 ORDER BY account_online.timestamp DESC
 LIMIT :param_2) AND account_online.account_id = :param_1
```
The SELECT statement itself does not look very deep, certainly not deep enough to trigger this warning. But then if we take a look at `statement._compile_options._current_path`, it has 12 things it. I don't understand why there are all of those `Mapper` objects in the path that have nothing to do with the SELECT statement being executed?
```
>>> print('\n'.join(str(x) for x in statement._compile_options._current_path))
Mapper[MediaInstance(media_instance)]
MediaInstance.contents
Mapper[ContentMedia(content_media)]
ContentMedia.content
Mapper[AccountContent(account_content)]
AccountContent.account
Mapper[Account(account)]
Account.entity
Mapper[Entity(entity)]
Entity.accounts
Mapper[Account(account)]
Account.last_online

>>> print(len(statement._compile_options._current_path))
12
```


To be clear, I am not doing any deeply-nested loader options at the point that this query execution is being triggered. This is the section of my code where the stack trace begins that ultimately ends in the warning:
```python
last_online_dates = [
    account.last_online.timestamp.astimezone(tzlocal()).date()
    for account in accounts
    if account.last_online
]
```
So, it is simply accessing the `account.last_online` relationship on an Account object. It should not be pulling in all of those other Mapper classes that are listed in `_current_path`, should it?

**Reply by zzzeek** [reply]
that _current_path thing matches up the options that were present when that account object was first loaded.  that's because if you said  options(selectinload(A.b).selectinload(B.c).selectinload(C.account).lazyload(Account.last_online).selectinload(Online.things).joinedload(Things.foo)), that last "lazyload()" sets itself up in terms of that path, so that it knows how to match up with the subsequent selectinload().joinedload()  path on there.  the whole thing gets preserved.

**Reply by dkbarn** [reply]
Could you explain what you mean "options that were present when that account object was first loaded"? If I had that many nested loader options when the account was first loaded, why wasn't the SAWarning raised then? Why is it being raised later on when an unrelated query is being executed? Is it because the cache was empty on first load, and this check only happens when the cache is not empty?

Do you have any tips for how to find where in my code the too-deep query ultimately responsible for this warning is, since the warning gives no indication and isn't actually being raised by the query at fault?

**Reply by CaselIT** [reply]
How are you loading accounts in your example?

Note that it's not an arbitrary query that sets up the patch, it's connectedness to the objects that trigger a lazy load, so you should look at how they are loaded

**Comment by zzzeek** [comment]
> Could you explain what you mean "options that were present when that account object was first loaded"? If I had that many nested loader options when the account was first loaded, why wasn't the SAWarning raised then? 

because the query stopped at the lazyload and the depth wouldn't have gotten that deep...

> Why is it being raised later on when an unrelated query is being executed? 

in the scenario I presented, which might not be what's actually happening for you, that is a "related" query.  if I sent options to lazyload(A.b).joinedload(B.c), then later, when I have an A object and I access "b", the joinedload() option all the way back from when it was first loaded is still in effect.   these queries are linked when this kind of thing happens.

> Do you have any tips for how to find where in my code the too-deep query ultimately responsible for this warning is, since the warning gives no indication and isn't actually being raised by the query at fault?

I think it's better that we create an issue for #10579, add some kind of option so this can be set explicitly, and then maybe in 2.1 or whatever I just take this entire warning out and instead have the SQL logging indicate the reason caching was disabled in this case.

**Reply by dkbarn** [reply]
Ok, thanks for taking the time to explain all this to me! I will keep an eye out for a new issue for #10579 and watch that.
